src/Model/Scope/BasicCondition.php

```diff
@@ -0,0 +1,340 @@
+    {
+        return array_filter([$this->key, $this->operator, $this->value]) ? false : true;
+    }
+
+    public function clear()
+    {
+        $this->key = $this->operator = $this->value = null;
+
+        return $this;
+    }
+
+    public function negate()
+    {
+        if ($this->operator && isset(self::$opposites[$this->operator])) {
+            $this->operator = self::$opposites[$this->operator];
+        } else {
+            throw new Exception('Negation of condition is not supported for ' . ($this->operator ?: 'no') . ' operator');
+        }
+
+        return $this;
+    }
+
+    public function toWords(bool $asHtml = false): string
+    {
+        if (!$this->getModel()) {
+            throw new Exception('Condition must be associated with Model to convert to words');
+        }
+
+        // make sure clones are used to avoid changes
+        $condition = clone $this;
+
+        $key = $condition->keyToWords($asHtml);
+
+        $operator = $condition->operatorToWords($asHtml);
+
+        $value = $condition->valueToWords($condition->value, $asHtml);
+
+        $ret = trim("{$key} {$operator} {$value}");
+
+        return $asHtml ? $ret : html_entity_decode($ret);
+    }
+
+    protected function keyToWords(bool $asHtml = false): string
+    {
+        $model = $this->getModel();
+
+        $words = [];
+
+        if (is_string($field = $this->key)) {
+            if (str_contains($field, '/')) {
+                $references = explode('/', $field);
+
+                $words[] = $model->getModelCaption();
+
+                $field = array_pop($references);
+
+                foreach ($references as $link) {
+                    $words[] = "that has reference {$this->readableCaption($link)}";
+
+                    $model = $model->refLink($link);
+                }
+
+                $words[] = 'where';
+
+                if ($field === '#') {
+                    $words[] = $this->operator ? 'number of records' : 'any referenced record exists';
+                    $field = '';
+                } elseif ($field === '?') {
+                    $words[] = 'any referenced record exists';
+                    $field = '';
+                } elseif ($field === '!') {
+                    $words[] = 'no referenced records exist';
+                    $field = '';
+                }
+            }
+
+            $field = $model->hasField($field) ? $model->getField($field) : null;
+        }
+
+        if ($field instanceof Field) {
+            $words[] = $field->getCaption();
+        } elseif ($field instanceof Expression) {
+            $words[] = "expression '{$field->getDebugQuery()}'";
+        }
+
+        $string = implode(' ', array_filter($words));
+
+        return $asHtml ? "<strong>{$string}</strong>" : $string;
+    }
+
+    protected function operatorToWords(bool $asHtml = false): string
+    {
+        return $this->operator ? (self::$dictionary[strtoupper((string) $this->operator)] ?? 'is equal to') : '';
```

Comment:
this should fail, operator if operator should be `=`, then is must be always set

Reply:
No, because operator may be `NULL` if the condition is based on `Expression` (raw SQL)

Reply:
I formulated it badly, there should be no `is equal to` fallback

Reply:
Better now?